### PR TITLE
API refactoring

### PR DESCRIPTION
**Pull request by kneth**
## What, How & Why?

This closes https://jira.mongodb.org/browse/RJS-689

* [x] Move `Realm.Sync` to `Realm.App.Sync`
* [x] Add `Realm.User.identities`
* [x] Add `Realm.App.Sync.syncSession`
* [x] Add `Realm.App.Sync.allSyncSessions`
* [x] Add `Realm.User.deviceId`
* [x] Add `Realm.User.providerType`
* [x] Rename `Realm.User.remoteMongoClient` to `Realm.User.mongoClient`
* [x] Rename `Realm.Credentials.custom` to `Realm.Credentials.jwt`
* [x] Add `Realm.MongoDB.Client.serviceName`
* [x] Add `Realm.MongoDB.Collection.name`
* [x] Add `Realm.MongoDB.Database.name`
* [ ] TypeScript definitions for `MongoClient`, etc.
* [ ] Add methods and properties to RN debugger support

## ☑️ ToDos

* [x] 📝 Changelog entry
* ~[ ] 📝 `Compatibility` label is updated or copied from previous entry~
* [x] 🚦 Tests
* ~[ ] 📝 Public documentation PR created or is not necessary~
* [x] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [ ] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: tests/js/open-behavior-tests.js [hunk 1]
```diff
@@ -156,7 +156,7 @@ module.exports = {
                 otherUserRealm.write(() => {
                     otherUserRealm.create(schemas.TestObject.name, { doubleCol: 42.133 });
                 });
-                return otherUserRealm.syncSession.uploadAllLocalChanges().then(() => {
+                return otherUserRealm.App.SyncSession.uploadAllLocalChanges().then(() => {
```
Comment (steffenagger) [inline tests/js/open-behavior-tests.js:159]: Isn't `uploadAllLocalChanges` an instance function, called on a class here?
(If so, this is done in more places)

**Review by steffenagger — COMMENTED** [review]
In general I would have preferred to see `Realm.Sync.User` moved to `Realm.App.User`, instead of `Realm.User` (as I believe it's App related).

It also confuses me a bit (as a new user...) to see both `Realm.App.SyncSession` AND `Realm.App.Sync.Session`.

File: lib/mongo_client.js [hunk 2]
```diff
@@ -41,6 +41,12 @@ class RemoteMongoDBCollection {
         this.serviceName = serviceName;
     }
 
+    /** @inheritdoc */
+    get name(
+    ) {
```
[the remark this answers is not in the document]
Comment (kneth) [reply lib/mongo_client.js:46]: I was trying to make it look the other methods - but they have arguments.

File: src/js_sync.hpp [hunk 3]
```diff
@@ -645,6 +649,64 @@ void SyncClass<T>::initialize_sync_manager(ContextType ctx, ObjectType this_obje
...
+        }
+        else if (Value::is_number(ctx, partition_value_value)) {
+            auto pv = Value::validated_to_number(ctx, partition_value_value);
+            partition_bson = bson::Bson(static_cast<int64_t>(pv));
+        }
+        else if (Value::is_object_id(ctx, partition_value_value)) {
+            auto pv = Value::validated_to_object_id(ctx, partition_value_value);
+            partition_bson = bson::Bson(pv);
+        }
+        else {
```
[the remark this answers is not in the document]
Comment (kneth) [reply src/js_sync.hpp:677]: Can only agree!

File: tests/js/open-behavior-tests.js (hunk 1 above)
Comment (kneth) [reply tests/js/open-behavior-tests.js:159]: It is a bad search/replace - good catch!

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,7 +4,18 @@ NOTE: Support for syncing with realm.cloud.io and/or Realm Object Server has bee
...
+* Renamed `Realm.User.remoteMongoClient()` to `Realm.User.mongoClient()`. ([RJS-689](https://jira.mongodb.org/browse/RJS-689))
+* Removed `Remote` prefix from all `MongoDB` related classes.
+* Renamed `Realm.Sync` to `Realm.App.Sync` including all methods and properties. ([RJS-689](https://jira.mongodb.org/browse/RJS-689))
+
 ### Enhancements
+* Added property `Realm.User.deviceId`. ([RJS-689](https://jira.mongodb.org/browse/RJS-689))
+* Added property `Realm.User.providerType`. ([RJS-689](https://jira.mongodb.org/browse/RJS-689))
+* Added property `Realm.User.identities`. ([RJS-689](https://jira.mongodb.org/browse/RJS-689))
+* Added property `name` to `Realm.MongoDB`, `Realm.MongoDB.Database` and `Realm.MongoDB.Collection`. ([RJS-689](https://jira.mongodb.org/browse/RJS-689))
+* Added methods `Realm.App.Sync.syncSession()` and `Realm.App.Sync.allSyncSessions()`. ([RJS-689](https://jira.mongodb.org/browse/RJS-689))
```
[the remark this answers is not in the document]
Comment (kneth) [reply CHANGELOG.md:18]: It is a good suggestion.

File: src/js_sync.hpp [hunk 5]
```diff
@@ -645,6 +684,45 @@ void SyncClass<T>::initialize_sync_manager(ContextType ctx, ObjectType this_obje
...
+template<typename T>
+void SyncClass<T>::get_sync_session(ContextType ctx, ObjectType this_object, Arguments &args, ReturnValue &return_value) {
+    args.validate_count(2);
+
+    auto user_object = Value::validated_to_object(ctx, args[0], "user");
+    SharedUser user = *get_internal<T, UserClass<T>>(ctx, user_object);
+
+    auto partition_value_value = args[1];
+    if (Value::is_undefined(ctx, partition_value_value)) {
+        return_value.set(Value::from_null(ctx));
```
Comment (blagoev) [inline src/js_sync.hpp:696]: could use return_value.set_null()

File: src/js_sync.hpp [hunk 6]
```diff
@@ -645,6 +684,45 @@ void SyncClass<T>::initialize_sync_manager(ContextType ctx, ObjectType this_obje
...
+    else {
+        std::string partition_value = partition_value_bson_to_string<T>(ctx, partition_value_value);
+
+        auto sync_config = SyncConfig(user, partition_value);
+        auto path = SyncManager::shared().path_for_realm(sync_config);
+        if (auto session = user->session_for_on_disk_path(path)) {
+            return_value.set(create_object<T, SessionClass<T>>(ctx, new WeakSession(session)));
+        }
+        else {
+            return_value.set(Value::from_null(ctx));
```
Comment (blagoev) [inline src/js_sync.hpp:707]: could use return_value.set_null()

File: src/js_sync.hpp [hunk 7]
```diff
@@ -645,6 +684,45 @@ void SyncClass<T>::initialize_sync_manager(ContextType ctx, ObjectType this_obje
...
+    args.validate_count(2);
+
+    auto user_object = Value::validated_to_object(ctx, args[0], "user");
+    SharedUser user = *get_internal<T, UserClass<T>>(ctx, user_object);
+
+    auto partition_value_value = args[1];
+    if (Value::is_undefined(ctx, partition_value_value)) {
+        return_value.set(Value::from_null(ctx));
+    }
+    else {
```
Comment (blagoev) [inline src/js_sync.hpp:698]: could remove the else here

File: src/js_sync.hpp [hunk 8]
```diff
@@ -645,6 +684,45 @@ void SyncClass<T>::initialize_sync_manager(ContextType ctx, ObjectType this_obje
...
+        std::string partition_value = partition_value_bson_to_string<T>(ctx, partition_value_value);
+
+        auto sync_config = SyncConfig(user, partition_value);
+        auto path = SyncManager::shared().path_for_realm(sync_config);
+        if (auto session = user->session_for_on_disk_path(path)) {
+            return_value.set(create_object<T, SessionClass<T>>(ctx, new WeakSession(session)));
+        }
+        else {
```
Comment (blagoev) [inline src/js_sync.hpp:706]: could remove else here

File: src/js_user.hpp [hunk 9]
```diff
@@ -212,6 +218,37 @@ void UserClass<T>::get_id(ContextType ctx, ObjectType object, ReturnValue &retur
...
+    return_value.set(Object::create_array(ctx, identity_objects));
+}
+
+template<typename T>
+void UserClass<T>::get_device_id(ContextType ctx, ObjectType object, ReturnValue &return_value) {
+    auto user = get_internal<T, UserClass<T>>(ctx, object)->get();
+    if (user->has_device_id()) {
+        return_value.set(user->device_id());
+    }
+    else {
```
Comment (blagoev) [inline src/js_user.hpp:241]: could remove else

**Review by blagoev — APPROVED** [review]
with minor comments

**Pull request merged**